packages/@aws-cdk/aws-apigateway/lib/authorizers/lambda.ts

```diff
@@ -0,0 +1,139 @@
+   *
+   * @default - no regex filter will be applied.
+   */
+  readonly validationRegex?: string;
+
+  /**
+   * An optional IAM role for APIGateway to assume before calling the Lambda-based authorizer. The IAM role must be
+   * assumable by 'apigateway.amazonaws.com'.
+   *
+   * @default - A resource policy is added to the Lambda function allowing apigateway.amazonaws.com to invoke the function.
+   */
+  readonly assumeRole?: iam.IRole;
+}
+
+/**
+ * Token based lambda authorizer that recognizes the caller's identity as a bearer token,
+ * such as a JSON Web Token (JWT) or an OAuth token.
+ * Based on the token, authorization is performed by a lambda function.
+ *
+ * @resource AWS::ApiGateway::Authorizer
+ */
+export class TokenAuthorizer extends AuthorizerBase {
+
+  /**
+   * The id of the authorizer.
+   * @attribute
+   */
+  public readonly authorizerId: string;
+
+  /**
+   * The ARN of the authorizer to be used in permission policies, such as IAM and resource-based grants.
+   * @attribute
+   */
+  public readonly authorizerArn: string;
+
+  constructor(scope: Construct, id: string, props: TokenAuthorizerProps) {
+    super(scope, id, {
+      physicalName: props.authorizerName || Lazy.stringValue({ produce: () => this.node.uniqueId })
+    });
+
+    if (props.resultsCacheTtl && props.resultsCacheTtl.toSeconds() > 3600) {
+      throw new Error(`Lambda authorizer property 'cacheTtl' must not be greater than 3600 seconds (1 hour)`);
+    }
+
+    const resource = new CfnAuthorizer(this, 'Resource', {
+      name: this.physicalName,
+      restApiId: this.restApiId,
+      type: 'TOKEN',
+      authorizerUri: `arn:aws:apigateway:${Stack.of(this).region}:lambda:path/2015-03-31/functions/${props.handler.functionArn}/invocations`,
+      authorizerCredentials: props.assumeRole ? props.assumeRole.roleArn : undefined,
+      authorizerResultTtlInSeconds: props.resultsCacheTtl && props.resultsCacheTtl.toSeconds(),
+      identitySource: props.identitySource || 'method.request.header.Authorizer',
+      identityValidationExpression: props.validationRegex,
+    });
+
+    this.authorizerId = super.getResourceNameAttribute(resource.ref);
+
+    this.authorizerArn = Stack.of(this).formatArn({
+      service: 'execute-api',
+      resource: this.restApiId,
+      resourceName: `authorizers/${this.authorizerId}`
+    });
+
+    if (!props.assumeRole) {
+      props.handler.addPermission(`${this.node.uniqueId}:Permissions`, {
+        principal: new iam.ServicePrincipal('apigateway.amazonaws.com'),
+        sourceArn: this.authorizerArn
+      });
+    } else if (props.assumeRole instanceof iam.Role) { // i.e., not imported
+      props.assumeRole.attachInlinePolicy(new iam.Policy(this, 'authorizerInvokePolicy', {
+        statements: [
+          new iam.PolicyStatement({
+            resources: [ props.handler.functionArn ],
+            actions: [ 'lambda:InvokeFunction' ],
+          })
+        ]
+      }));
+    }
+  }
```

Comment:
if assumeRole is defined, don't you still need lambda permission for the service principal?

Reply:
Isn't that specified at and only at the time of construction of `iam.Role`, like [here](https://github.com/aws/aws-cdk/blob/nija-at/apigateway-lambdaauthorizer/packages/%40aws-cdk/aws-apigateway/test/authorizers/integ.token-authorizer-iam-role.ts#L16-L18)?

Something else I'm missing?

Reply:
You should be able to add a policy statement to a role.

Reply:
Sorry, I meant: don’t you need to also include the lambda permission for the service principal? Just do a small integ test that validates that both options work.

Reply:
Yep, already have an integ test. The link above points to that.